### ISSUE TITLE
Change detection not run when the errors are set on reactive forms

### ISSUE THREAD

**Issue by Vikki123** [angular/angular#29515]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/forms....

### Is this a regression?

Not sure.

### Description

I am on angular 7 and i am using reactive forms for one of my forms and the code is as below. (Stackblitz [Here][1])

**Template**

     <form [formGroup]="SignupForm" *ngIf="SignupForm">
       <input type="text" class="form-control" id="username" formControlName = "username">
       <div *ngIf="SignupForm.errors?.errors">Error</div>
     </form>

**Typescript**

    export class AppComponent implements OnInit {
      SignupForm: FormGroup;
    
      constructor() {
    
      }
    
      ngOnInit() {
        setTimeout(() => {
          this.SignupForm = new FormGroup({
            'username': new FormControl('Value'),
            'email': new FormControl()
          });
          this.SignupForm.setErrors({errors: 'Error'});
        })
      }
    }

As you can see i am trying to set the error as soon as the form is initialized (In my actual case i would do some validation based on back end data and then do this. ). But the error doesn't seem to show up even though `SignupForm.errors` is set with the error object.

But if i do 

    setTimeout(() => { this.SignupForm.setErrors({errors: 'Error'});}) 

Everything works fine. Am i doing something wrong or is it an angular bug which i can report. Please help. Thank you.

  [1]: https://stackblitz.com/edit/angular-forms-bug?file=src%2Fapp%2Fapp.component.ts

**Use case**

Actually the thing is my form comes from the backend which is asynchronous. After the i create the form group from the metadata i do the validation and set the errors. I have used setTimeout on ngOnInit just to depict its asynchronous.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-forms-bug?file=src%2Fapp%2Fapp.component.ts

## 🌍  Your Environment

**Angular Version:**
"@angular/common": "^7.2.5",
    "@angular/compiler": "^7.2.5",
    "@angular/core": "^7.2.5",
    "@angular/forms": "^7.2.5",
    "@angular/platform-browser": "^7.2.5",
    "@angular/platform-browser-dynamic": "^7.2.5",
    "@angular/router": "^7.2.5",
    "core-js": "^2.6.5",
    "rxjs": "^6.4.0",
    "zone.js": "^0.8.29"

**Comment by kara** [comment]
@Vikki123 This is expected behavior, though I can see why it would be confusing in this case.  It has to do with when change detection runs.

When you set the `this.SignupForm` property in `ngOnInit`, it's passed into the `FormGroupDirective` input. Whenever that value changes, `FormGroupDirective` input must update the value/validity of the form to match whatever `FormGroup` is passed in. This happens when change detections runs, which is after the synchronous call to `setErrors`.  So you essentially manually override the errors generated by validation, then run validation again (which correctly resets the errors based on validator results).  

tldr: To manually override the errors, generally you need to wait until all the forms have had a chance to update themselves. Instead of `setTimeout`, you could also try listening to the first statusChange event on the form and set the errors then.

**Comment by Vikki123** [comment]
Hi @kara  thanks for the reply. But listening to status change will not be useful in this case i guess. Because my code is as follows

```
ngOnInit() {
    this.form = this.service.getFormGroup(tool.Forms); // Creates the form group dynamically,.
    // Now i have to check if the validation of the form based on some metadata.
}
```
If i understand correctly, statusChanges will be triggered after the user interacts with the form. My problem is that i need to set the errors before the user even interacts with the form. I understand that its by design that its not possible. Is there any cleaner way of doing this. If i have misunderstood something with respect to statusChanges event please correct me. An example would be really helpful. Thanks .
